### ISSUE TITLE
parsing of anisotropic B factors in  Bio/PDB/MMCIFParser.py uses wrong mmCIF category

### ISSUE THREAD

**Issue by berrisfordjohn** [biopython/biopython#1733]
In Bio/PDB/MMCIFParser.py

The mmCIF parser tries to parse aniso records from the mmCIF file. 
It uses _atom_site.aniso_U[1][1] to determine if an atom has aniso records. 

aniso records are separate from atom_site in mmCIF
It should use the mmCIF category _atom_site_anisotrop

See (for example)
https://www.ebi.ac.uk/pdbe/entry-files/3nir.cif

Current behavior - all structures will have aniso_flag of 0

The tests for comparing the PDB and mmCIF parsers seem to miss this.

**Comment by svalqui** [comment]
>>> from Bio.PDB.MMCIFParser import MMCIFParser
>>> parser = MMCIFParser()
>>> structure = parser.get_structure("3nir.cif", "/home/s/Downloads/3nir.cif")
/home/s/PythonProjects/biopython/Bio/PDB/StructureBuilder.py:127: PDBConstructionWarning: WARNING: Residue (' ', 22, ' ') redefined at line 434.
  PDBConstructionWarning)
/home/s/PythonProjects/biopython/Bio/PDB/StructureBuilder.py:127: PDBConstructionWarning: WARNING: Residue (' ', 25, ' ') redefined at line 490.
  PDBConstructionWarning)
>>> ato = structure[0]["A"][1]["N"]
>>> coo = ato.get_coord()
>>> print(coo)
[  3.265 -14.107  16.877]
>>> ani = ato.get_anisou()
>>> print(ani)
None
>>> 

I'll have a look at this, I should add further comments on 4-6 weeks.

**Comment by berrisfordjohn** [comment]
If you have to change the mmCIF parsing you might consider using an external mmCIF parser.
At PDBe (pdbe.org) we use a pure python parser: https://gitlab.com/glenveegee/PDBeCIF
or there is a joint project between CCP4 and Global Phasing to produce an mmCIF parser
https://github.com/project-gemmi/gemmi
I use the later to update mmCIF files https://github.com/berrisfordjohn/adding_stats_to_mmcif

**Comment by jgreener64** [comment]
The goal of the Biopython mmCIF parser is slightly different from others: to extract a hierarchical protein representation rather than just add/remove/change mmCIF dictionary entries.

mmCIF parsing has two steps in Biopython - reading the file into a dictionary-style data structure and then assembling a Biopython structure object from that.

Whilst we can always learn a lot from other implementations, most of the mmCIF issues raised (including this one) deal with the second step. I think the first parsing step is pretty robust. At least, a number of people use it and it seems to work well.

**Comment by rtf-const** [comment]
```
>>> from Bio.PDB.MMCIFParser import MMCIFParser
>>> parser = MMCIFParser()
>>> structure = parser.get_structure("3nir.cif", "/home/const/Downloads/3nir.cif")
/usr/local/lib/python3.6/dist-packages/Bio/PDB/StructureBuilder.py:127: PDBConstructionWarning: WARNING: Residue (' ', 22, ' ') redefined at line 434.
  PDBConstructionWarning)
/usr/local/lib/python3.6/dist-packages/Bio/PDB/StructureBuilder.py:127: PDBConstructionWarning: WARNING: Residue (' ', 25, ' ') redefined at line 490.
  PDBConstructionWarning)
>>> ato = structure[0]["A"][1]["N"]
>>> coo = ato.get_coord()
>>> print(coo)
[  3.265 -14.107  16.877]
>>> ani = ato.get_anisou()
>>> print(ani)
[ 0.0704 -0.02    0.0103  0.0658  0.0029  0.0428]
```

**Comment by svalqui** [comment]
Replaced _atom_site.aniso_U with _atom_site_anisotrop.U, and created an index by _atom_site_anisotrop.id, so it can be related back to _atom_site.id.

@peterjc  Hi Peter here an amendment that address the issue raised, I have tested it with 3nir.cif that I found on the PDB web site; also works with 4CUP.cif on /Test/PDB as below

>>> from Bio.PDB.MMCIFParser import MMCIFParser
>>> parser = MMCIFParser()
>>> structure = parser.get_structure("4CUP.cif", "/home/s/PythonProjects/biopython/Tests/PDB/4CUP.cif")
>>> for atom in structure.get_atoms():
...     print("a_name:", atom.get_name(), " a_id: ", atom.get_id(), " a_anisou: ", atom.get_anisou())
... 

@berrisfordjohn  Hi John thank you for risen the issue and include the details, I think the functionally on the parser fits Biopython purposes, there considerable work already put on it so I have updated it to address this issue.

Considerations:
1. PDB documentation shows that anisotropic matrix U values can be found on either here ATOM_SITE or here ATOM_SITE_ANISOTROP; one or the other but both, they are mutual exclusive; at this stage the amendment is only for _atom_site_anisotrop.U, given time constrains its only narrow it to address this issue does not consider _atom_site.aniso_U

Regards,

Sergio.

**Comment by peterjc** [comment]
Fixed via #1775, see also #1822 for any followup...

**Comment by andrei-istrate** [comment]
In this discussion it was mentioned that it was "created an index by _atom_site_anisotrop.id, so it can be related back to _atom_site.id". 

However in the current implementation the relation between _atom_site_anisotrop.id and _atom_site.id is not checked which means that the _atom_site_anisotrop.U values are correctly assigned only if _atom_site_anisotrop.id and _atom_site.id follow the same order and there are no missing _atom_site_anisotrop records.

An  error is reported only if there are _atom_site_anisotrop records for exactly one atom:

from Bio.PDB.MMCIFParser import MMCIFParser
parser = MMCIFParser(QUIET=True)
structure = parser.get_structure('6v9q.cif', '6v9q.cif')

Traceback (most recent call last):
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/playground/test.py", line 11, in <module>
    structure = parser.get_structure('name', path)
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/venv/lib/python3.6/site-packages/Bio/PDB/MMCIFParser.py", line 64, in get_structure
    self._build_structure(structure_id)
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/venv/lib/python3.6/site-packages/Bio/PDB/MMCIFParser.py", line 203, in _build_structure
    mapped_anisou = [float(_) for _ in u]
  File "/Users/andrei/PycharmProjects/3D-STRUDEL/venv/lib/python3.6/site-packages/Bio/PDB/MMCIFParser.py", line 203, in <listcomp>
    mapped_anisou = [float(_) for _ in u]
ValueError: could not convert string to float: '.'

**Comment by svalqui** [comment]
@andrei-istrate  Yes, I had created an index on #1822 you can see the code there, but at the time #1775 presented a more complete work and addressed the immediate issue #1733 ; also I couldn't find the file that highlighted the issue, So the index is not implemented.
Could you please open a new issue with the details above and I'll review and give more details about this.
